storage: return term_id in timequery_result

nvartolomei opened the pull request:
Main commit:

    storage: return term_id in timequery_result

    This is required for fixing a ListOffsets bug. The bug fix will be
    implemented later for all ListOffsets calls rather than just for
    timequeries.

    Ref: https://redpandadata.atlassian.net/browse/CORE-12505

Not backporting for now. Will look at backporting everything at once needed for fixing the actual bug to avoid dealing with conflicts multiple times.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
storage/tests: compare timequery_result rather than specific fields

Less error prone and forces developer to update tests when the result
changes. I want to add new fields to the result and I did not like the
fact that no tests broke when the result shape changed.

nvartolomei committed:
storage: return term_id in timequery_result

This is required for fixing a ListOffsets bug. The bug fix will be
implemented later for all ListOffsets calls rather than just for
timequeries.

Ref: https://redpandadata.atlassian.net/browse/CORE-12505

nvartolomei committed:
storage/tests: also use compressed batches in timequery tests

This is supported not from long ago but the test wasn't updated.

nvartolomei committed:
storage/tests: prepare for testing timequery with custom term

nvartolomei committed:
storage/tests: test timequery with different terms

nvartolomei committed:
treewide: untangle some includes

WillemKauf reviewed (dismissed):
> co_return timequery_result{b.term(), result_o, result_t};

LGTM

WillemKauf commented:
```
/root/.cache/bazel/_bazel_root/28dca059ead9e3cd868100b70ad4bca7/sandbox/linux-sandbox/3036/execroot/_main/src/v/storage/log_reader.cc:589:32: error: 'b' used after it was moved [bugprone-use-after-move,-warnings-as-errors]
  589 |     co_return timequery_result{b.term(), result_o, result_t};
      |                                ^
/root/.cache/bazel/_bazel_root/28dca059ead9e3cd868100b70ad4bca7/sandbox/linux-sandbox/3036/execroot/_main/src/v/storage/log_reader.cc:571:27: note: move occurred here
  571 |     auto batch = co_await internal::decompress_batch(std::move(b));
      |    
```

nvartolomei commented:
If you also wonder why it is not highlighted by clangd (i.e. in IDE) https://github.com/llvm/llvm-project/blob/ed2bfd132509da679320a1d691af4a91192297d0/clang-tools-extra/clangd/TidyProvider.cpp#L217-L220

nvartolomei committed:
storage: use the same variable for decompressed batch

Keeping both b and batch is scope makes it easy to use-after-move.
True story: https://github.com/redpanda-data/redpanda/pull/26919#issuecomment-3097082040

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1afd91f2-2070-47cd-ae20-c9374de71475-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69363">build#69363</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest    | test_rack_awareness  |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69363#01982de0-6f2b-444c-ac12-d85ca9ed791f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69363#01982de0-6f33-4b44-b764-dc10dc54602c | FLAKY         | 18/21    | upstream reliability is '99.57446808510639'. current run reliability is '85.71428571428571'. drift is 13.86018 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->
